### DISCUSSION TITLE
PGroonga 2.4.2 リリース!

### DISCUSSION THREAD

**Discussion by komainu8** [pgroonga/pgroonga#272] [Releases]
## PGroonga（ぴーじーるんが） 2.4.2リリース

PostgreSQLで高速日本語全文検索をできるようにする[PGroonga](https://pgroonga.github.io/ja/)の2.4.2をリリースしました！

### 改良

  * [``pgroonga_wal_apply()``](https://pgroonga.github.io/ja/reference/functions/pgroonga-wal-apply.html) と [``pgroonga_vacuum()``](https://pgroonga.github.io/ja/reference/functions/pgroonga-vacuum.html) をスタンバイデータベース上で自動的に実行する新しいモジュール [``pgroonga_standby_maintainer``](https://pgroonga.github.io/ja/reference/modules/pgroonga-standby-maintainer.html) を追加しました。

  * [pgroonga_snippet_html()](https://pgroonga.github.io/ja/reference/functions/pgroonga-snippet-html.html) スニペットの長さを指定する新しい引数を追加しました。[GitHub#253](https://github.com/pgroonga/pgroonga/issues/253) [Reported by askdkc], [GitHub#255](https://github.com/pgroonga/pgroonga/pull/255) [Patched by askdkc]

  * Windows版のPOstgreSQL 15をサポートしました。 [GitHub#256](https://github.com/pgroonga/pgroonga/issues/256) [Reported by Raif Atef]

  * Ubuntu 22.04のPostgreSQL 15をサポートしました。

### 修正

  * 2フェーズコミット中のVACUUMが失敗する問題を修正しました。 [GitHub#252](https://github.com/pgroonga/pgroonga/issues/252) [Reported by Raif Atef]

### ``pgroonga_wal_apply()`` と ``pgroonga_vacuum()`` をスタンバイデータベース上で自動的に実行する新しいモジュール ``pgroonga_standby_maintainer`` を追加しました

  通常、PGroongaとストリーミングレプリケーションを一緒に使っている場合、PGroongaのWALはスタンバイのデータベースには適用されません。

  したがって、例えば、スタンバイのデータベースをプライマリーのデータベースに切り替えた後の最初の"SELECT"は遅い可能性があります。 切り替わった後のプライマリーのデータベースに未適用のPGroongaのWALを適用する時間があるためです。

  いままでは、この問題は [``pgroonga_wal_applier`` モジュール](https://pgroonga.github.io/ja/reference/modules/pgroonga-wal-applier.html)を使うことで解決できましたが、2.4.2からは ``pgroonga_standby_maintainer`` モジュールを使うことをおすすめします。
  ``pgroonga_standby_maintainer`` モジュールは、``pgroonga_wal_applier`` モジュールよりも便利です。
  ``pgroonga_standby_maintainer`` モジュールは、 ``pgroonga_wal_apply()`` の自動実行だけでなく、 ``pgroonga_vacuum()`` 関数の自動実行もおこなうためです。

  したがって、pgroonga_standby_maintainerモジュールを使えば、スタンバイのデータベース上で、 pgroonga_wal_applier モジュールを使う必要も pgroonga_vacuum() 関数 を定期的に実行する必要もありません。

  ``pgroonga_standby_maintainer`` モジュールの使い方については、 [pgroonga_standby_maintainer モジュールのリファレンス](https://pgroonga.github.io/ja/reference/modules/pgroonga-standby-maintainer.html)も参照してください。

### [pgroonga_snippet_html()] スニペットの長さを指定する新しい引数を追加しました

  通常、 ``pgroonga_snippet_html`` 関数で抽出するキーワードの周辺テキストの最大数は200バイトですが、この機能を使うことでこの最大数を調整できます。
  そのため、より柔軟にキーワードの周辺テキストを抽出できます。

  例えば、キーワードの周辺テキストを50バイト抽出したい場合は、以下のように実行できます。
  ``pgroonga_snippet_html()`` の第三引数によってキーワードの周辺テキストの長さを動的に指定できることに注目してください。

  ```sql
  SELECT pgroonga_snippet_html(
    'Groonga is a fast and accurate full text search engine based on ' ||
    'inverted index. One of the characteristics of Groonga is that a ' ||
    'newly registered document instantly appears in search results. ' ||
    'Also, Groonga allows updates without read locks. These characteristics ' ||
    'result in superior performance on real-time applications.' ||
    E'\n' ||
    E'\n' ||
    'Groonga is also a column-oriented database management system (DBMS). ' ||
    'Compared with well-known row-oriented systems, such as MySQL and ' ||
    'PostgreSQL, column-oriented systems are more suited for aggregate ' ||
    'queries. Due to this advantage, Groonga can cover weakness of ' ||
    'row-oriented systems.',
    ARRAY['Groonga'],
    50
  );
                                                                                                                    pgroonga_snippet_html                                                                                                                     
  ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  {"<span class=\"keyword\">Groonga</span> is a fast and accurate full text search en","he characteristics of <span class=\"keyword\">Groonga</span> is that a newly regi","search results. Also, <span class=\"keyword\">Groonga</span> allows updates witho"}
  (1 row)
  ```

  ``pgroonga_snippet_html()`` 関数の使い方については [pgroonga_snippet_html()のリファレンス](https://pgroonga.github.io/ja/reference/functions/pgroonga-snippet-html.html)も参照してください。

## まとめ

新規ユーザーの方は、[PGroongaについて](https://groonga.org/ja/blog/2017/10/10/pgroonga-2.0.2.html#pgroonga)も参照してください。

PostgreSQLで高速に日本語全文検索をしたいという方はPGroongaを使ってガンガン検索してください！

**Comment by askdkc** [comment]
@komainu8 さん

つかぬことをお聞きしますがmacOSの[brew](https://formulae.brew.sh/formula/pgroonga)への最新版反映のタイミングって、ここでの新リリースから通常どの程度かかる感じなのでしょうか？

**Reply by kou** [reply]
PGroongaプロジェクトとして更新していないのでだれかがやったら更新になります！
自分で`brew bump-formula-pr pgroonga`で更新することもできます！
https://docs.brew.sh/Manpage#bump-formula-pr-options-formula

`bump-formula-pr`するとなんかいい感じに動いてPRを作るところまでやってくれます。

**Reply by askdkc** [reply]
雰囲気でやってみました👀

[https://github.com/Homebrew/homebrew-core/pull/117049](https://github.com/Homebrew/homebrew-core/pull/117049)

**Reply by askdkc** [reply]
[https://formulae.brew.sh/formula/pgroonga#default](https://formulae.brew.sh/formula/pgroonga#default)

2.4.2になったっぽい👀

<img height="400" src="https://user-images.githubusercontent.com/7894265/205030538-abf57ddb-9d1d-4bf8-ac53-77f7c25ca254.png">

**Reply by kou** [reply]
Homebrewの人たちって反応速いんですよ！

**Reply by askdkc** [reply]
> Homebrewの人たちって反応速いんですよ！

ものの数時間でアップデート処理に回されてますものね👀
最早脊髄反射🏎

**Reply by kou** [reply]
`bump-formula-pr`で作られたやつはCIが通っていればすぐにマージされることがほとんどです！

そうじゃなくて、手動でちょろちょろいじっているやつはたまにマージされるまでにやいのやいのすることがあります。

**Reply by askdkc** [reply]
> 手動でちょろちょろいじっているやつはたまにマージされるまでにやいのやいのすることがあります。

https://qiita.com/cosmo0920/items/ed7e071d111c533e217c

例えばこの辺の話みたいなパターンですか？
何か大変そう。。。

**Reply by kou** [reply]
いやぁ、これは関係ないですね。

https://github.com/Homebrew/homebrew-core/pull/114034 とかです。

**Reply by askdkc** [reply]
なるほど！
勉強になります！
